drivers/peci/peci_handlers.c

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2020 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <drivers/peci.h>
+#include <syscall_handler.h>
+
+static inline int z_vrfy_peci_config(struct device *dev, u32_t bitrate)
+{
+	Z_OOPS(Z_SYSCALL_DRIVER_PECI(dev, config));
+
+	return z_impl_peci_config((struct device *)dev, (u16_t)bitrate);
+}
+#include <syscalls/peci_config_mrsh.c>
+
+static inline int z_vrfy_peci_enable(struct device *dev)
+{
+	Z_OOPS(Z_SYSCALL_DRIVER_PECI(dev, enable));
+
+	return z_impl_peci_enable((struct device *)dev);
+}
+#include <syscalls/peci_enable_mrsh.c>
+
+static inline int z_vrfy_peci_disable(struct device *dev)
+{
+	Z_OOPS(Z_SYSCALL_DRIVER_PECI(dev, disable));
+
+	return z_impl_peci_disable((struct device *)dev);
+}
+#include <syscalls/peci_disable_mrsh.c>
+
+static inline int z_vrfy_peci_transfer(struct device *dev,
+				       const struct peci_msg *msg)
+{
+	struct peci_msg msg_copy;
```

Comment:
This still isn't correct, and unfortunately we're not close.
Did the peci_msg data structure change since the last time I reviewed this? I swear I remember this being simpler last time I looked.

Right now, in the current version of this PR we have:

```
struct peci_msg
     -----> struct peci_buf (tx_buffer)
            ----> buf
            ----> fcs
     -----> struct peci_buf (rx_buffer)
            ----> buf
            ----> fcs
```

So basically the user is sending over **seven** different memory regions which have to be validated. The containing peci_msg, two linked peci_buf, each containing a buffer pointer and an fcs pointer.

Why did you split this up in this way? Can you please consider collapsing this down to a smaller number of memory regions?

The problem is that all of these memory regions have to be considered potentially malicious. Not only that, but they cannot be validated in-place if they contain any information that must also be validated, a copy must be made. Otherwise, a malicious user could simply modify these data structures after the kernel has validated them.

So unless something changes, you are going to have to:

1. Make a validated copy of msg into msg_copy on the stack with z_user_from_copy()
2. Make validated copies of msg->tx_buffer and msg->rx_buffer into tx_buffer_copy and rx_buffer_copy onto the stack, again using z_user_from_copy(), and update msg_copy.tx_buffer and msg_copy.rx_buffer to point to those instead of the original values.
3. Validate the buf/size for each copied peci_buf
4. Validate the fcs member for each copied peci_buf. How is this even validatated? Is it NULL terminated like a C string? Is it the same size as buf? How do we prove that the user has access to this pointer?

Reply:
@andrewboie Implemented all check in peci_handler as per instructions. 

struct peci_buf is based on **struct spi_buf** from **spi.h** which encapsulates a buffer. Main difference with these spi and i2c structures is the peci driver returns the checksum in single byte hence the fcs parameter. I git rid of fcs field and document this byte is part of rx buffer data.

Regarding further collapse, could you elaborate? I modeled  **struct peci_msg** based on **struct spi_buf_set**  and **struct i2c_msg**, however instead the structure holding an array of structs or buffers, it only has 2 "buffer" structs. Is the array of structs preferred to multiple members?